Use batch processing in edit-widgets package

adamziel opened the pull request:
This PR updates the `edit-widgets` package so that it saves the entities using batch processing against the correct API endpoints.

Testing instructions TBD

adamziel committed:
Draft batch-processing module

adamziel committed:
Draft batch-processing module

TimothyBJacobs commented:
This looks great to me!

> Chunks processing seems to be one step too far as they make the entire transaction not atomic.

Could you expand what this means?

adamziel committed:
committing/rudimentary API_FETCH integration

adamziel committed:
committing/rudimentary API_FETCH integration

adamziel committed:
Make batch processing functional!

adamziel committed:
Make batch processing functional!

adamziel committed:
Roll back integration changes

adamziel committed:
Roll back integration changes

adamziel committed:
Cleanup package.json

adamziel committed:
Cleanup package.json

adamziel committed:
Update package-lock.json

adamziel committed:
Update package-lock.json

adamziel commented:
>> Chunks processing seems to be one step too far as they make the entire transaction not atomic.
> Could you expand what this means?

@TimothyBJacobs sure! The words "transaction" and "commit" usually characterize an atomic operation that either goes through or fails in its entirety. In this PR, however, a transaction may consists of multiple chunks which means it may partially fail so I'm not entirely comfortable with this vocabulary.

I'm not saying we shouldn't do chunks - that's a useful automation. If the caller needs to save 21 entities and a single batch may only handle 20, then it's going to send two batch requests anyway. Handling that logic manually would be repetitive and prone to errors so it's good to have it baked in. And there transactions out there that may end up in a partially failed state too (MyISAM - I'm looking at you).

I think I would be more comfortable with this if the chunks were called "transactions" and transactions were called something else. This way each transaction would be atomic, but the collection of transactions wouldn't.

adamziel committed:
Update terminology

adamziel committed:
Update terminology

adamziel committed:
yield from commitTransaction

adamziel committed:
yield from commitTransaction

adamziel committed:
Correct yield from

adamziel committed:
Correct yield from

adamziel commented:
There we go, I updated the nomenclature to "batch" and "transaction".

adamziel committed:
Move the comment where it belongs

adamziel committed:
Move the comment where it belongs

TimothyBJacobs committed:
First pass at enqueueItemAndWaitForResults.

TimothyBJacobs commented:
One issue I'm running into is that the `batch` we return to the caller takes the ordered list of items in a queue, and returns them as an object. Object insertion order isn't guaranteed. This means that if the caller doesn't have the `itemId` returned by `enqueueItem`, which in nearly all cases it won't. They can't match up their `apiFetch` calls with the results returned by `processBatch`.

Keying the `results` and `errors` by the `itemId` makes a lot of operations convenient, so maybe instead of reverting that to an array, we maintain a property on the batch of the `itemId`s in insertion order?

youknowriad commented:
@adamziel would you mind explaining the public API of the proposed package in the description or the README?

adamziel committed:
update package-lock.json

TimothyBJacobs committed:
First pass at using the new sidebars and widget endpoints. (#26086)

* Draft batch-processing module

* committing/rudimentary API_FETCH integration

* Make batch processing functional!

* Roll back integration changes

* Cleanup package.json

* Update package-lock.json

* Update terminology

* yield from commitTransaction

* Correct yield from

* Move the comment where it belongs

* First pass at enqueueItemAndWaitForResults.

* First pass at using the new sidebars and widget endpoints.

* Add test for saving multiple widgets in a row

* Generate better error message when creating fails

* Add basic batch integration.

Works off of #26205.

* Allow mixing HTTP methods in a batch.

* Fix disappearing data, add basic error handling.

We need to get the full list of widget ids from the client, we can't use the
returned widget ids from the batch because not all widgets will be dirty.

Also implements basic error handling for when the batch fails. A snackbar is
added with the names of the widgets that failed to save. We also now propagate
a dummy error to the apiFetch callers for the requests that didn't have a
validation error, but need something to short circuit their handling.

* yield from the widget area saving.

* Move batch-processing into edit-widgets package

Co-authored-by: Adam Zieliński <adam@adamziel.com>

adamziel committed:
remove separate batch-processing package

adamziel commented on packages/edit-widgets/src/store/batch-processing/reducer.js:
```diff
@@ -0,0 +1,221 @@
+			};
+		}
+
+		case 'SETUP_PROMISE': {
+			return {
+				...state,
+				promises: {
+					...state.promises,
+					[ action.queue ]: {
+						...( state.promises[ action.queue ] || {} ),
+						[ action.context ]: {
+							promise: action.promise,
+							resolve: action.resolve,
+							reject: action.reject,
+						},
+					},
+				},
+			};
+		}
```
Promises are mutable but redux state should be immutable by definition - let's remove this in a follow up PR.

TimothyBJacobs replied:
So putting this in just a global somewhere?

adamziel replied:
I was thinking more about not storing this anywhere, only returning it to the caller. It's not a very well-thought proposal, I understand you added that for a reason. Let's figure it out after 9.2 RC1 - this works and is fine for now.

adamziel committed:
Update batch processing tests

adamziel committed:
Resolve batch errors in a safe, well-ordered way

adamziel committed:
Fix unit tests

adamziel committed:
Wire progress indicator to specific widgets

adamziel committed:
Fix authorship after merge/rebase (1/2)

adamziel committed:
Fix authorship after merge/rebase (2/2)

TimothyBJacobs committed:
Merge branch 'update/data-layer-batch-requests-use-state' of github.com:WordPress/gutenberg into update/data-layer-batch-requests-use-state

TimothyBJacobs committed:
Don't mark the method property as required.

We don't evaluate nested defaults yet, so this caused errors.

TimothyBJacobs committed:
Allow updating widgets without including the id_base or number.

adamziel committed:
Don't start GET /wp/v2/widgets?per_page=-1 if there are batch requests already in progress

adamziel committed:
Process newWidgetClientIds in correct order

adamziel committed:
Fix package.json

adamziel committed:
remove batch-processing package
